pyiron_base/project/delayed.py

```diff
@@ -0,0 +1,273 @@
+import hashlib
+import inspect
+import re
+from typing import Optional, Tuple
+
+import cloudpickle
+
+
+def get_function_parameter_dict(funct: callable) -> dict:
+    """
+    Get dictionary of parameters for a function
+
+    Args:
+        funct (callable): the function to get the parameters for
+
+    Returns:
+        dict: parameters for the function
+    """
+    return {
+        k: None if v.default == inspect._empty else v.default
+        for k, v in inspect.signature(funct).parameters.items()
+    }
+
+
+def get_hash(binary: bytes) -> str:
+    """
+    Get the hash of a binary string - remove the specification of jupyter kernel from hash to be deterministic
+
+    Args:
+        binary (bytes): binary string to hash
+
+    Returns:
+        str: hash of the binary string
+    """
+    binary_no_ipykernel = re.sub(b"(?<=/ipykernel_)(.*)(?=/)", b"", binary)
+    return str(hashlib.md5(binary_no_ipykernel).hexdigest())
+
+
+def get_graph(
+    obj: object,
+    obj_name: str = None,
+    nodes_dict: dict = {},
+    edges_lst: list = [],
+    link_node: Optional[str] = None,
+) -> Tuple[dict, list]:
+    """
+    Get dictionary of nodes with node names as keys and node objects as values. In addition, generate a list of edges,
+    consisting of pairs of node names which are linked together.
+
+    Args:
+        obj (object): Object to generate dictionary of nodes and list of edges
+        obj_name (str): Name of the object
+        nodes_dict (dict): Dictionary of nodes
+        edges_lst (list): List of edges
+        link_node (str): Name of the node to link to
+
+    Returns:
+        dict, list: dictionary of nodes and list of edges
+    """
+    if isinstance(obj, DelayedObject) and obj_name is None:
+        try:
+            obj_name = obj._function.__name__
+        except TypeError:
+            obj_name = str(obj).replace("<", "").replace(" object at ", "")
+    if obj_name is None:
+        try:
+            obj_name = obj.__name__
+        except AttributeError:
+            obj_name = str(type(obj))
+    try:
+        node_name = obj_name + "_" + get_hash(binary=cloudpickle.dumps(obj))
+    except TypeError:
+        node_name = obj_name
+    nodes_dict.update({node_name: obj})
```

Comment:
This should be a separate function. `get_graph` is too long otherwise.

Reply:
It is now separate in an `get_node_name()` function.